src/libcore/tests/iter.rs

```diff
@@ -1549,6 +1549,51 @@ fn test_repeat_take_collect() {
     assert_eq!(v, vec![42, 42, 42]);
 }
 
+#[test]
+fn test_repeat_with() {
+    #[derive(PartialEq, Debug)]
+    struct NotClone(usize);
+    let mut it = repeat_with(|| NotClone(42));
+    assert_eq!(it.next(), Some(NotClone(42)));
+    assert_eq!(it.next(), Some(NotClone(42)));
+    assert_eq!(it.next(), Some(NotClone(42)));
+    assert_eq!(repeat_with(|| NotClone(42)).size_hint(), (usize::MAX, None));
+}
+
+#[test]
+fn test_repeat_with_rev() {
```

Comment:
This behaviour is surprising to me, since it produces the same sequence of different things with or without the `.rev()`.  Perhaps it should only be DEI for F:Fn, not F:FnMut?

Reply:
Yeah I agree. `F: Fn() -> A` seems reasonable under the assumption that it should be a pure function, and we can always document that on `repeat_with` (the function).

Reply:
It's consistent with map this way... No big opinion but I'd vote to keep it double ended with FnMut

Reply:
@bluss I guess that works too if we document it well... I don't have a big opinion either.